### PR TITLE
feat: Add gradient filters to match ASE's FrechetCellFilter

### PR DESCRIPTION
**Pull request by n-gao**
The goal of this PR is to match ASE's joint coordinate and cell optimization. For that, we needed a few mechanics that were introduced in #201 #202 and this PR. #201 and #202 enable us to use arbitrary (not just lower triangular) lattice vectors and define frame based on reference frames.

In this PR we change update the interface for gradient computations in our make_xyz_potential functions to take filters that allow us to specify dependencies between variables. This is currently only available for position and cells.

With these changes we match ASE 1:1 in the relaxation of a MOF with UMA.
image
image

## Summary of changes

Introduces a unified `Geometry`/`PositionsAndCell` relaxation filter API in `kups/potential/filter.py`, replacing the scattered `compute_position_and_cell_gradients: bool` flag across all potential constructors with an explicit `gradient: Lens[Geometry, PositionsAndCell] | None` parameter.

Three concrete filters are provided:

- `POSITIONS_AND_CELL` — raw positions + cell DOFs (atoms pinned, partial `∂E/∂h|_r`)
- `POSITIONS_ONLY` — positions only; the cell rides along stop-gradiented so `∂E/∂cell` is zero while the DOF pytree shape stays compatible with cell filters
- `FRECHET_FILTER` — atoms-ride-the-cell exp-map parameterisation matching ASE's `FrechetCellFilter`; DOFs are `(q, A)` with `q = r @ h⁻¹ @ R` and `r = q @ R⁻¹ @ h` on set, so the virial coupling falls out of autodiff

`make_relax_propagator` drops the `optimize_cell: bool` flag; callers now pass the same filter they built the potential with. The propagator's `property` lens is composed from a `to_geometry` helper and the caller-supplied filter, so the atoms-ride-the-cell coupling is applied on every optimizer `set` rather than being reconstructed post-hoc via `total_lattice_gradient`.

For direct (torch) MLIAP potentials a `filter_pullback` function is added in `mliap/direct.py`. It maps the model's physical `(∂E/∂r, ∂E/∂h|_r)` cotangents to `∂E/∂u` for any filter DOF pytree `U` via a single `jax.vjp` through `filter.set`, so the chain rule (including the Fréchet exp-map adjoint and the atoms-ride-the-cell term) is handled automatically. `torch_mliap_model_fn` is simplified to always emit `PositionsAndCell`; the pullback is applied in `make_torch_mliap_potential` when a filter is provided.

The `lattice_gradient_from_virial` recovery in the torch interface is corrected: the position virial is subtracted without symmetrisation before solving for `∂E/∂h`, recovering the full (including antisymmetric) lattice gradient rather than only the symmetric part.

The L-BFGS state (`ScaleByAseLbfgsState`) is refactored to store parameter and update history as flat lists of raw array leaves rather than pytrees, eliminating repeated `Table` validation inside `jax.lax.scan` and avoiding multiple donations of shared `Index` buffers. A `treedef` field is added to unflatten the preconditioned update on exit.

The convergence check in `run_relax` is updated to compare `max|∂E/∂u|` (the cached DOF gradient) against the force tolerance directly, replacing the previous mixed force/cell-gradient check that used `total_lattice_gradient`.

New tests cover filter lens laws, Fréchet transform parity with `scipy.expm_frechet`, step-for-step agreement with ASE's `UnitCellFilter` and `FrechetCellFilter` on shear-free and sheared triclinic cells, `filter_pullback` correctness for all three filters, and `lattice_gradient_from_virial` recovery under she

### CODE REVIEW

**Comment by n-gao** [comment]
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#215](https://github.com/cusp-ai-oss/kUPS/pull/215) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/application/potential/filter.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def _positions_only_set(g: Geometry, value: PositionsAndCell) -> Geometry:
+    particles = g.particles.map_data(
+        lambda p: PositionsAndSystemIndex(value.positions.data, p.system)
+    )
+    cell = jax.lax.stop_gradient(value.cell.data)
+    return Geometry(particles, g.systems.set_data(cell))
+
+
```
Comment (felixmusil) [inline src/kups/application/potential/filter.py:37]: missing the `CELL_ONLY` option

**Review by felixmusil — APPROVED** [review]
please add the comment to the TODO list (non-blocking here)

**Comment by n-gao** [comment]
### Merge activity

* **Jul 1, 3:57 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204).
